bootstore/src/db/mod.rs

```diff
@@ -76,41 +67,95 @@ impl Db {
+        conn: &mut SqliteConnection,
+        rack_uuid: &Uuid,
         epoch: i32,
-        share: SerializableShareDistribution,
+        share_distribution: SerializableShareDistribution,
     ) -> Result<(), Error> {
-        info!(self.log, "Writing key share prepare for {epoch} to the Db");
+        assert_ne!(0, epoch);
         use schema::key_shares::dsl;
-        // We save the digest so we don't have to deserialize and recompute most of the time.
-        // We'd only want to do that for a consistency check occasionally.
-        let val = bcs::to_bytes(&share)?;
-        let share_digest =
-            sprockets_common::Sha3_256Digest(Sha3_256::digest(&val).into())
-                .into();
-        let prepare = KeyShare {
-            epoch,
-            share: Share(share),
-            share_digest,
-            committed: false,
-        };
-        diesel::insert_into(dsl::key_shares)
-            .values(&prepare)
-            .execute(&mut self.conn)?;
-        Ok(())
+        let prepare = KeyShare::new(epoch, share_distribution)?;
+        conn.immediate_transaction(|tx| {
+            // Has the rack been initialized?
+            if !self.is_initialized(tx)? {
+                return Err(Error::RackNotInitialized);
+            }
+
+            // Does the rack_uuid match what's stored?
+            self.validate_rack_uuid(tx, rack_uuid)?;
+
+            // Check for idempotence
+            if let Some(stored_key_share) = dsl::key_shares
+                .filter(dsl::epoch.eq(epoch))
+                .get_result::<KeyShare>(tx)
+                .optional()?
+            {
+                if prepare == stored_key_share {
+                    return Ok(());
+                } else {
+                    return Err(Error::KeyShareAlreadyExists { epoch });
+                }
+            }
+
+            // We don't allow shares for old epochs
+            if let Some(stored_epoch) = dsl::key_shares
+                .select(dsl::epoch)
+                .filter(dsl::epoch.gt(epoch))
+                .get_result::<i32>(tx)
+                .optional()?
+            {
+                return Err(Error::OldKeySharePrepare { epoch, stored_epoch });
+            }
+
+            info!(
+                self.log,
+                "Writing key share prepare for epoch {epoch} to the Db"
+            );
+            diesel::insert_into(dsl::key_shares)
+                .values(&prepare)
+                .execute(tx)?;
+            Ok(())
+        })
     }
 
     pub fn commit_share(
-        &mut self,
+        &self,
+        conn: &mut SqliteConnection,
+        rack_uuid: &Uuid,
         epoch: i32,
         digest: sprockets_common::Sha3_256Digest,
     ) -> Result<(), Error> {
         use schema::key_shares::dsl;
-        self.conn.immediate_transaction(|tx| {
+        conn.immediate_transaction(|tx| {
+            // Does the rack_uuid match what's stored?
+            self.validate_rack_uuid(tx, rack_uuid)?;
+
```

Comment:
I know I wrote in the PR description that we should not accept a Commit if there is even a prepare for a later epoch. We don't do that here. I'm not actually sure if we should. For instance, is it possible that a node was prepared, offline for the commit, then somehow got a later prepare that included itself? That seems unlikely and maybe we can prevent it at the nexus layer. However, if we don't prevent it, we may still need the commit to allow the sled to unlock itself while a new prepare is in progress. I need to think about this some more.

Reply:
I"m leaning toward actually allowing this. I'm not sure it's harmful. I will continue to think about it as this protocol evolves.